tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/schema_util.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence, Tuple, TypeVar, Union, cast
+
+import attrs
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+
+"""
+Specification of DataFrame/Table schema. Uses motivating this (over a simple pyarrow.Schema):
+
+1. allow for late-binding to various types, e.g., Dictionary index_type (set to smallest width,
+based upon number of labels).
+
+2. allow explicit type choices where there are options (e.g., string vs large_string)
+
+"""
+
+OptDataFrame = TypeVar("OptDataFrame", pd.DataFrame, None, Optional[pd.DataFrame])
+
+
+@attrs.define(frozen=True, kw_only=True, slots=True)
+class FieldSpec:
```

Comment:
Even if this is not intended as a public class, adding some docstrings could help better understand some of the methods (and especially parameters) to first-time readers.

Reply:
done